ledger/participant-state/kvutils/tools/src/main/scala/ledger/participant/state/kvutils/tools/integritycheck/StateUpdateComparison.scala

```diff
@@ -162,6 +177,26 @@ object ReadServiceStateUpdateComparison {
       case u: Update.CommandRejected => u.copy(recordTime = newRecordTime)
     }
 
+  private def dropFetchAndLookupByKeyNodes[Nid, Cid](
+      tx: VersionedTransaction[Nid, Cid]
+  ): VersionedTransaction[Nid, Cid] = {
+    val nodes = tx.nodes.filter {
+      case (_, _: Node.NodeFetch[Cid] | _: Node.NodeLookupByKey[Cid]) => false
+      case _ => true
+    }
+    val filteredNodes = nodes.map {
+      case (nid, node: Node.NodeExercises[Nid, Cid]) =>
+        val filteredNode = node.copy(children = node.children.filter(nodes.contains))
+        (nid, filteredNode)
+      case keep => keep
+    }
+    val filteredRoots = tx.roots.filter(filteredNodes.contains)
+    val version = filteredRoots.iterator.foldLeft(TransactionVersion.minVersion)((acc, nodeId) =>
+      TransactionVersion.Ordering.max(acc, nodes(nodeId).version)
+    )
```

Comment:
Isn't there an utility for that? I'd rather avoid replicating this logic here.

Reply:
@remyhaemmerle-da Do you have a suggestion?

Reply:
This is KV specific.  I will rather let it here.

Reply:
If there isn't already encapsulated logic to decide a transaction's version based on its nodes, I'd at least extract it as a separate method for improved readability.

Reply:
Daml-LF code assumes there is **only** the engine and the deserialization that creates transaction object. 
The logic to infer the version is nested deeply inside the engine.

Reply:
The question is: do you update the version in KV when you rid the transaction of its Fetch and Lookup Nodes ?

Reply:
No, we don't update it, we just set it to the transaction's version (see [here](https://github.com/digital-asset/daml/blob/ecd3110f488566c3ceef3c19c023e8c2e81c960a/ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/transaction/TransactionCommitter.scala#L355)).